### PR TITLE
Document how functional collectives work under eager/dynamo

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104387
* __->__ #104386

Move user facing apis to the top for best visibility
(strictly code-motion in this PR, besides adding comments)

### CODE REVIEW

**Review by voznesenskym — APPROVED** [review]

**Comment by wconstab** [comment]
cc @kumpera @wanchaol

File: torch/distributed/_functional_collectives.py [hunk 1]
```diff
@@ -76,6 +76,221 @@ def is_torchdynamo_compiling():
...
+
+
+"""
+User facing APIs for functional collectives
+-------------------------------------------
+
+These apis are called by usercode and expected to work both in eager execution and compilation,
+but there are significant differences to how the two modes are implemented underneath.
+
+Eager execution is 'optimized' using a tensor subclass that schedules the synchronization (via wait_tensor() op)
```
Comment (wanchaol) [inline torch/distributed/_functional_collectives.py:95]: nit: we should also probably remove or merge the previous comments L33-36 to this one?

**Review by wanchaol — APPROVED** [review]
lgtm, one nit. Thanks for better documentations!

File: torch/distributed/_functional_collectives.py [hunk 2]
```diff
@@ -76,6 +76,221 @@ def is_torchdynamo_compiling():
...
+The different types will be desugared to a canonical format
+"""
+RANK_TYPES = Union[List[int], List[List[int]], dist.ProcessGroup, "dist._tensor.DeviceMesh", Tuple["dist._tensor.DeviceMesh", int]]
+
+
+"""
+User facing APIs for functional collectives
+-------------------------------------------
+
+These apis are called by usercode and expected to work both in eager execution and compilation,
```
Comment (fduwjj) [inline torch/distributed/_functional_collectives.py:92]: nit: user code?

Comment (wconstab) [reply torch/distributed/_functional_collectives.py:92]: fixing in next PR

**Pull request closed without merging**